cirq/ops/projectors.py

```diff
@@ -0,0 +1,98 @@
+from typing import Any, Dict, Iterable, List, Tuple, TYPE_CHECKING, Union
+
+import numpy as np
+
+from cirq import value
+from cirq.ops import raw_types
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@value.value_equality
+class Projector(raw_types.Gate):
+    """A non-unitary gate that projects onto one or more qubits.
+
+    The input is a matrix representing the basis of the space we project onto.
+    The basis vectors need not be orthogonal, but they must be independent (i.e.
+    the matrix has full rank).
+
+    For example, if you want to project on |0>, you would provide the basis
+    [[1, 0]]. To project onto |10>, you would provide the basis [[0, 0, 1, 0]].
+    If you want to project on the space spanned by |10> and |11>, you could
+    provide the basis [[0, 0, 1, 0], [0, 0, 0, 1]].
+    """
+
+    def __init__(self,
+                 projection_basis: Union[List[List[float]], np.ndarray],
+                 qid_shape: Tuple[int, ...] = (2,),
+                 enfore_orthogonal_basis: bool = False):
```

Comment:
nit: Since you're actually enforcing orthonormality, not just orthogonality (IOW, you're enforcing orthogonality and unit norm), this should be named "enforce_orthonormal_basis".

Reply:
Done.